### PR TITLE
Create SD Card Module and Activity Store/Load

### LINKED ISSUE #37: Create SD Card Module

### PR DESCRIPTION
**Pull request by jennaluz**

### CODE REVIEW

File: partitions.csv [hunk 1]
```diff
@@ -2,5 +2,5 @@
 # Note: if you have increased the bootloader size, make sure to update the offsets to avoid overlap
 nvs,      data, nvs,     0x9000,  0x6000,
 phy_init, data, phy,     0xf000,  0x1000,
-factory,  app,  factory, 0x10000, 1M,
+factory,  app,  factory, 0x10000, 0x16E360,
```
Comment (loganfin) [inline partitions.csv]: Please resolve the merge conflict

File: src/lumen/activity/football.cpp [hunk 2]
```diff
@@ -58,9 +65,97 @@ void Football::button_pressed(ButtonEvent event)
         team_two_.score().decrease();
         break;
     default:
-        ESP_LOGE("activity/football", "Unknown button event");
+        ESP_LOGE(tag, "Unknown button event");
         break;
     }
 }
 
+void Football::load(json data)
```
Comment (loganfin) [inline src/lumen/activity/football.cpp]: The parameter on this and all of the other Activity::load functions should be `json const& data`.

File: src/lumen/hardware/sd_card.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+/** Write a JSON object to a file.
+ *
+ * \param filepath The path to the file to write to.
+ *
+ * \param data The JSON object to write.
+ *
+ * \note `filepath` must be prepended by '/'
+ */
+void write_json(std::string filepath, nlohmann::json data);
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.hpp]: This should also be a const&

File: src/lumen/hardware/sd_card.hpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+/** Delete a file.
+ *
+ * \param filepath The path to the file to write to.
+ *
+ * \note `filepath` must be prepended by '/'
+ */
+void delete_file(std::string filepath);
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.hpp]: This should also be a const&

File: src/lumen/hardware/sd_card.hpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+/** Read a file stream into a JSON object.
+ *
+ * \param filepath The path to the file to read from.
+ *
+ * \returns The file as a JSON object.
+ *
+ * \note `filepath` must be prepended by '/'
+ */
+nlohmann::json read_json(std::string filepath);
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.hpp]: This should be a const&

File: src/lumen/hardware/sd_card.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
+#include "lumen/hardware/sd_card.hpp"
+
+#include "esp_vfs_fat.h"
+#include "sdmmc_cmd.h"
+
+#include "driver/gpio.h"
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Why is this separated the two above includes?

File: src/lumen/hardware/sd_card.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: `<iostream>` isn't used

File: src/lumen/hardware/sd_card.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+using json = nlohmann::json;
+
+namespace lumen::hardware {
+
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Remove the line break here

File: src/lumen/hardware/sd_card.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+namespace {
+
+constexpr auto tag = "hardware";
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Should this be "hardware/sd_card"

File: src/lumen/hardware/sd_card.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    ESP_LOGI(tag, "Filesystem mounted");
+    initialized = true;
+    sdmmc_card_print_info(stdout, card);
+}
+
+json read_json(std::string filepath)
+{
+    std::ifstream file;
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Check if the SD card was initialized.

File: src/lumen/hardware/sd_card.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+    json data;
+    file >> data;
+
+    return data;
+}
+
+void write_json(std::string filepath, json data)
+{
+    std::ofstream file;
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Check if the SD card was initialized.

File: src/lumen/hardware/sd_card.cpp [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        ESP_LOGE(tag, "Unable to write to file '%s'", filepath.c_str());
+        return;
+    }
+
+    file << data;
+}
+
+void delete_file(std::string filepath)
+{
+    std::filesystem::remove(sd_card_mount_point + filepath);
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp:145]: Check if the SD card was initialized.

File: src/lumen/hardware/sd_card.cpp [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    // Configure the SD card slot without chip detection.
+
+    sdspi_device_config_t slot_config = SDSPI_DEVICE_CONFIG_DEFAULT();
+    slot_config.host_id = host_slot;
+    slot_config.gpio_cs =
+        static_cast<gpio_num_t>(CONFIG_HARDWARE_SD_CARD_CS_PIN);
+
+    // Configure and mount the SD card.
+
+    sdmmc_card_t* card;
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Initialize this

File: src/lumen/hardware/sd_card.cpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    esp_vfs_fat_sdmmc_mount_config_t mount_config = {
+        .format_if_mount_failed = true,
+        .max_files = 5,
+        .allocation_unit_size = 16 * 102
+    };
+
+    ESP_LOGI(tag, "Mounting filesystem");
+
+    ret = esp_vfs_fat_sdspi_mount(
+        static_cast<char const*>(sd_card_mount_point),
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: You shouldn't need to cast this

File: src/lumen/hardware/sd_card.cpp [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        &card
+    );
+
+    if (ret != ESP_OK) {
+        if (ret == ESP_FAIL) {
+            ESP_LOGE(
+                tag,
+                "Failed to mount filesystem. "
+                "If you want the card to be formatted, set the "
+                "CONFIG_EXAMPLE_FORMAT_IF_MOUNT_FAILED menuconfig option."
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: Remove this

File: src/lumen/activity/activity.hpp [hunk 16]
```diff
@@ -34,6 +36,18 @@ class Activity {
      */
     virtual void button_pressed(ButtonEvent event);
 
+    /** Load a saved activity.
+     *
+     * \param data The JSON object with the saved activity data.
+     */
+    virtual void load(nlohmann::json data);
```
Comment (loganfin) [inline src/lumen/activity/activity.hpp]: This should be const&

**Review by loganfin — CHANGES_REQUESTED** [review]

File: src/lumen/hardware/sd_card.hpp (hunk 3 above)
Comment (jennaluz) [reply src/lumen/hardware/sd_card.hpp]: I believe that the `json` data will be optimized. You can refer to the use case in `Context.store_activity()`. I made `filepath` a `std::string const&`.

File: src/lumen/hardware/sd_card.cpp [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        return {};
+    }
+
+    json data;
+    file >> data;
+
+    return data;
+}
+
+void write_json(std::string const& filepath, json data)
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.cpp]: `const&` parameters can take r-value arguments.

File: src/lumen/activity/activity.cpp [hunk 18]
```diff
@@ -1,15 +1,32 @@
...
     return display_.get();
 }
 
-void Activity::button_pressed(ButtonEvent event)
+void Activity::button_pressed(ButtonEvent /* event */)
+{
+    ESP_LOGE(tag, "Button handler is not implemented");
+}
+
+void Activity::load(json& /* data */)
```
Comment (loganfin) [inline src/lumen/activity/activity.cpp]: `const&`

File: src/lumen/activity/activity.hpp [hunk 19]
```diff
@@ -34,6 +36,18 @@ class Activity {
      */
     virtual void button_pressed(ButtonEvent event);
 
+    /** Load a saved activity.
+     *
+     * \param data The JSON object with the saved activity data.
+     */
+    virtual void load(nlohmann::json& data);
```
Comment (loganfin) [inline src/lumen/activity/activity.hpp]: `const&`

File: src/lumen/hardware/sd_card.hpp [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+/** Write a JSON object to a file.
+ *
+ * \param filepath The path to the file to write to.
+ *
+ * \param data The JSON object to write.
+ *
+ * \note `filepath` must be prepended by '/'
+ */
+void write_json(std::string const& filepath, nlohmann::json data);
```
Comment (loganfin) [inline src/lumen/hardware/sd_card.hpp]: `nlohmann::json const& data`

File: src/lumen/activity/football.cpp [hunk 21]
```diff
@@ -55,9 +62,115 @@ void Football::button_pressed(ButtonEvent event)
         team_two_.score().decrease();
         break;
     default:
-        ESP_LOGE("activity/football", "Unknown button event");
+        ESP_LOGE(tag, "Unknown button event");
         break;
     }
 }
 
+void Football::load(json& data)
```
Comment (loganfin) [inline src/lumen/activity/football.cpp]: `const&`

File: src/lumen/activity/football.hpp [hunk 22]
```diff
@@ -38,6 +40,18 @@ class Football : public Activity {
      */
     void button_pressed(ButtonEvent event) override;
 
+    /** Load a saved activity.
+     *
+     * \param data The JSON object with the saved football data.
+     */
+    void load(nlohmann::json& data) override;
```
Comment (loganfin) [inline src/lumen/activity/football.hpp]: `const&`

**Review by loganfin — CHANGES_REQUESTED** [review]

**Review by loganfin — APPROVED** [review]

**Pull request merged**